metadata/src/validation.rs

```diff
@@ -0,0 +1,851 @@
+) -> [u8; 32] {
+    match ty_def {
+        TypeDef::Composite(composite) => {
+            let mut bytes = hash(&[TypeBeingHashed::Composite as u8]);
+            for field in &composite.fields {
+                bytes = hash_hashes(bytes, get_field_hash(registry, field, visited_ids));
+            }
+            bytes
+        }
+        TypeDef::Variant(variant) => {
+            let mut bytes = hash(&[TypeBeingHashed::Variant as u8]);
+            for var in &variant.variants {
+                bytes = hash_hashes(bytes, get_variant_hash(registry, var, visited_ids));
+            }
+            bytes
+        }
+        TypeDef::Sequence(sequence) => {
+            let bytes = hash(&[TypeBeingHashed::Sequence as u8]);
+            xor(
+                bytes,
+                get_type_hash(registry, sequence.type_param.id, visited_ids),
+            )
+        }
+        TypeDef::Array(array) => {
+            // Take length into account; different length must lead to different hash.
+            let len_bytes = array.len.to_be_bytes();
+            let bytes = hash(&[
+                TypeBeingHashed::Array as u8,
+                len_bytes[0],
+                len_bytes[1],
+                len_bytes[2],
+                len_bytes[3],
+            ]);
+            xor(
+                bytes,
+                get_type_hash(registry, array.type_param.id, visited_ids),
+            )
+        }
+        TypeDef::Tuple(tuple) => {
+            let mut bytes = hash(&[TypeBeingHashed::Tuple as u8]);
+            for field in &tuple.fields {
+                bytes = hash_hashes(bytes, get_type_hash(registry, field.id, visited_ids));
+            }
+            bytes
+        }
+        TypeDef::Primitive(primitive) => {
+            // Cloning the 'primitive' type should essentially be a copy.
+            hash(&[TypeBeingHashed::Primitive as u8, primitive.clone() as u8])
+        }
+        TypeDef::Compact(compact) => {
+            let bytes = hash(&[TypeBeingHashed::Compact as u8]);
+            xor(
+                bytes,
+                get_type_hash(registry, compact.type_param.id, visited_ids),
+            )
+        }
+        TypeDef::BitSequence(bitseq) => {
+            let mut bytes = hash(&[TypeBeingHashed::BitSequence as u8]);
+            bytes = xor(
+                bytes,
+                get_type_hash(registry, bitseq.bit_order_type.id, visited_ids),
+            );
+            bytes = xor(
+                bytes,
+                get_type_hash(registry, bitseq.bit_store_type.id, visited_ids),
+            );
+            bytes
+        }
+    }
+}
+
+/// Obtain the hash representation of a `scale_info::Type` identified by id.
+fn get_type_hash(registry: &PortableRegistry, id: u32, visited_ids: &mut HashSet<u32>) -> [u8; 32] {
+    // Guard against recursive types and return a fixed arbitrary hash
+    if !visited_ids.insert(id) {
+        return hash(&[123u8]);
+    }
+
+    let ty = registry.resolve(id).unwrap();
```

Comment:
expect if this is infallible plz :)

Reply:
Yep that makes sense, we call this function internally and get the type ID from the metadata which should have the type registered in the registry. Have changed it to an `expect`